src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryCopy.java

```diff
@@ -0,0 +1,696 @@
+    public static void copyToArray(
+            MemorySegment srcSegment, long srcOffsetBytes,
+            byte[] dstArray, int dstIndexBytes, int dstCopyLengthBytes) {
+        Objects.requireNonNull(srcSegment);
+        Objects.requireNonNull(dstArray);
+        AbstractMemorySegmentImpl srcImpl = (AbstractMemorySegmentImpl)srcSegment;
+        srcImpl.checkAccess(srcOffsetBytes, dstCopyLengthBytes, true);
+        Objects.checkFromIndexSize(dstIndexBytes, dstCopyLengthBytes, dstArray.length);
+        scopedMemoryAccess.copyMemory(srcImpl.scope(), null,
+                srcImpl.unsafeGetBase(), srcImpl.unsafeGetOffset() + srcOffsetBytes,
+                dstArray, BYTE_BASE + dstIndexBytes, dstCopyLengthBytes);
+    }
+
+    //CHAR
+    /**
+     * Copies a number of char elements from a source char array to a destination segment,
+     * starting at a given array index, and a given segment offset (expressed in bytes).
+     * @param srcArray the source char array.
+     * @param srcIndexChars the starting index of the source char array.
+     * @param srcCopyLengthChars the number of char elements to be copied.
+     * @param dstSegment the destination segment.
+     * @param dstOffsetBytes the starting offset, in bytes, of the destination segment.
+     */
+    @ForceInline
+    public static void copyFromArray(
+            char[] srcArray, int srcIndexChars, int srcCopyLengthChars,
+            MemorySegment dstSegment, long dstOffsetBytes) {
+        copyFromArray(srcArray, srcIndexChars, srcCopyLengthChars, dstSegment,dstOffsetBytes, ByteOrder.nativeOrder());
+    }
+
+    /**
+     * Copies a number of char elements from a source char array to a destination segment,
+     * starting at a given array index, and a given segment offset (expressed in bytes), using the given byte order.
+     * @param srcArray the source char array.
+     * @param srcIndexChars the starting index of the source char array.
+     * @param srcCopyLengthChars the number of char elements to be copied.
+     * @param dstSegment the destination segment.
+     * @param dstOffsetBytes the starting offset, in bytes, of the destination segment.
+     * @param order the byte order to be used for the copy operation. If the specified byte order is
+     * different from the native order, a byte swap operation will be performed on each array element.
+     */
+    @ForceInline
+    public static void copyFromArray(
+            char[] srcArray, int srcIndexChars, int srcCopyLengthChars,
+            MemorySegment dstSegment, long dstOffsetBytes,
+            ByteOrder order) {
+        Objects.requireNonNull(srcArray);
+        Objects.requireNonNull(dstSegment);
+        Objects.checkFromIndexSize(srcIndexChars, srcCopyLengthChars, srcArray.length);
+        AbstractMemorySegmentImpl destImpl = (AbstractMemorySegmentImpl)dstSegment;
+        destImpl.checkAccess(dstOffsetBytes, srcCopyLengthChars << 1, false);
+        if (order == ByteOrder.nativeOrder()) {
```

Comment:
I like this code much more than the previous one, as it makes clear: swap tru or false, just a boolean! The Usage of MemorySegment.copyFrom() was very complicated as it used ValueLayouts, one for each size.